### PR TITLE
Extras custom reorg input validation

### PR DESCRIPTION
**Pull request by nishill**
I have finished adding the following requirements for validation:

1. check if the applet input can be found and the applet ID is valid
2. check if the applet has at least CONTRIBUTE access to the project
3. check if the file input can be found and the file ID is valid

After Quek's review I removed comments that are obvious, removed snake case, and fileDescribe declaration

### CODE REVIEW

File: src/test/scala/dxWDL/base/ExtrasTest.scala [hunk 1]
```diff
@@ -225,7 +229,8 @@ class ExtrasTest extends FlatSpec with Matchers {
 
     it should "throw IllegalArgumentException due to missing inputs in custom_reorg section" taggedAs(EdgeTest) in {
 
-        val app_id: String = "applet-123456"
+        // app_id is mummer nucmer app in project-FJ90qPj0jy8zYvVV9yz3F5gv
+        val app_id: String = "applet-FKfZpF002fp738k1Jqz8vX1G"
```
Comment (xquek) [inline src/test/scala/dxWDL/base/ExtrasTest.scala]: let change all to camelcase 
app_id -> appId

File: src/main/scala/dxWDL/base/Extras.scala [hunk 2]
```diff
@@ -618,11 +618,74 @@ object Extras {
...
+                  "Please set the value to null if there is no input."
             )
             case Some(x) => x
         }
 
+        return (reorgAppId,reorgInput)
+
+    }
+
+    def verifyReorgAppHasAccess(appDescribe: DXApplet.Describe, reorgAppId: String) = {
```
Comment (xquek) [inline src/main/scala/dxWDL/base/Extras.scala]: change to private method

File: src/main/scala/dxWDL/base/Extras.scala [hunk 3]
```diff
@@ -618,11 +618,74 @@ object Extras {
...
+        if ( access != "CONTRIBUTE" && access != "ADMINISTER" ) {
+
+            throw new PermissionDeniedException(s"ERROR: Applet for custom reorg stage ${reorgAppId } does not " +
+              s"have CONTRIBUTOR or ADMINISTRATOR access and this is required.", -1)
+
+        }
+
+    }
+
+    def veryifyRorgApp(reorgAppId: String) = {
```
Comment (xquek) [inline src/main/scala/dxWDL/base/Extras.scala]: private method

File: src/main/scala/dxWDL/base/Extras.scala [hunk 4]
```diff
@@ -618,11 +618,74 @@ object Extras {
...
+    def verifyInputs(reorgInput: String) = {
+
+        // if provided, check that the fileID is valid and present
+        if ( reorgInput != "" ) {
+            // format dx file ID
+            val reorgFileID: String = reorgInput.replace("dx://", "")
+            // if input file  ID is invalid, DXFile.getInstance will thown an IllegalArgumentException
+            val file: DXFile = DXFile.getInstance(reorgFileID)
+            // if reorgFileID cannot be found, describe will throw a ResourceNotFoundException
+            val fileDescribe: DXFile.Describe = file.describe()
```
Comment (xquek) [inline src/main/scala/dxWDL/base/Extras.scala]: No need to assign it to a value if it is never used.

File: src/main/scala/dxWDL/base/Extras.scala (hunk 2 above)
Comment (xquek) [reply src/main/scala/dxWDL/base/Extras.scala]: add Unit as the return type

File: src/main/scala/dxWDL/base/Extras.scala [hunk 5]
```diff
@@ -618,11 +618,74 @@ object Extras {
...
+        return (reorgAppId,reorgInput)
+
+    }
+
+    def verifyReorgAppHasAccess(appDescribe: DXApplet.Describe, reorgAppId: String) = {
+
+        // check applet has access to the projet
+        val accessJson = appDescribe.getAccess()
+        val projectAccess = accessJson.get("project")
+        val access: String = if ( projectAccess != null) {
```
Comment (xquek) [inline src/main/scala/dxWDL/base/Extras.scala]: consider using pattern matching -

```      
  val isValid: Boolean = accessJson.get("project") match {
            case null => false
            case x:JsonNode  => x.toString.replace("\"", "") match {
                case _ "CONTRIBUTE" | "ADMINISTER" => true
                case _ => false
            }
            case _ => false
        }

        if (!isValid) {
            throw new PermissionDeniedException(s"ERROR: Applet for custom reorg stage ${reorgAppId } does not " +
              s"have CONTRIBUTOR or ADMINISTRATOR access and this is required.", -1)

        }

File: src/main/java/com/dnanexus/DXApplet.java [hunk 6]
```diff
@@ -249,11 +249,22 @@ public Builder setTitle(String title) {
         private RunSpecification runSpec;
         @JsonProperty
         private String dxapi;
+        @JsonProperty
```
Comment (xquek) [inline src/main/java/com/dnanexus/DXApplet.java]: we will let ohad decide if he allows changes to be made on the dnanexus toolkit level.

File: src/main/scala/dxWDL/base/Extras.scala [hunk 7]
```diff
@@ -618,11 +618,74 @@ object Extras {
...
+
+    def parseCustomReorgAttrs(jsv: JsValue, verbose: Verbose): Option[ReorgAttrs] = {
+        if (jsv == JsNull)
+            return None
+
+        val fields = jsv.asJsObject.fields
+        // check required inputs are supplied
+        val (reorgAppId, reorgInput) = checkAttrs(fields)
+
+        // verify reorg app
```
[the remark this answers is not in the document]
Comment (nishill) [reply src/main/scala/dxWDL/base/Extras.scala]: done

File: src/test/scala/dxWDL/base/ExtrasTest.scala (hunk 1 above)
Comment (nishill) [reply src/test/scala/dxWDL/base/ExtrasTest.scala]: done

File: src/main/scala/dxWDL/base/Extras.scala (hunk 2 above)
Comment (nishill) [reply src/main/scala/dxWDL/base/Extras.scala]: done

File: src/main/scala/dxWDL/base/Extras.scala (hunk 3 above)
Comment (nishill) [reply src/main/scala/dxWDL/base/Extras.scala]: done

File: src/main/scala/dxWDL/base/Extras.scala (hunk 4 above)
Comment (nishill) [reply src/main/scala/dxWDL/base/Extras.scala]: done

File: src/main/scala/dxWDL/base/Extras.scala (hunk 5 above)
Comment (nishill) [reply src/main/scala/dxWDL/base/Extras.scala]: done

**Comment by nishill** [comment]
@orodeh could you please review now?

**Pull request merged**